### PR TITLE
Fixes #17, closes #16, closes #15, closes #14, initial refactoring for milestone 2.0

### LINKED ISSUE #14: Rewrite the help command to dynamically generate.
The []help command and the []modhelp command should be automatically generated from the command modules. Use Attributes to indicate the values that should be displayed.

### LINKED ISSUE #15: Rewrite the lookup tables.
The lookup tables are a bit sloppily coded at the moment, so give them a rewrite. Non-complex data structures like the rule table probably won't even need their own class.

### LINKED ISSUE #16: Get rid of static variables.
Right now, a lot of static variables are used for cross-class communication.
Fix this by applying dependancy injection.

### LINKED ISSUE #17: Change the leaderboard to use the nickname of a user.
Currently the username is displayed. Change this to display the nickname (if the user has one).

### PR DESCRIPTION
**Pull request by YilianSource**

### CODE REVIEW

File: Commands/HelpCommand.cs [hunk 1]
```diff
@@ -31,22 +34,52 @@ public async Task ModHelp ()
...
+        /// </summary>
         private EmbedBuilder GetHelpDialog(string mode)
         {
             EmbedBuilder eb = new EmbedBuilder()
                 .WithColor(new Color(247, 22, 131))
                 .WithTitle("BrackeysBot")
                 .WithDescription("The official Brackeys server bot. Commands are:");
 
-            string prefix = BrackeysBot.Configuration["prefix"];
+            string prefix = "[]";
```
Comment (CodeMyst) [inline Commands/HelpCommand.cs]: Use the prefix from the configuration file instead of hardcoding it in.

File: Commands/HelpCommand.cs [hunk 2]
```diff
@@ -31,22 +34,52 @@ public async Task ModHelp ()
...
+                eb.AddField(title, content);
             }
 
             return eb;
         }
+
+        /// <summary>
+        /// Gathers the data for all commands found in the command modules.
+        /// </summary>
+        private IEnumerable<HelpDataAttribute> GetCommandDatas (string mode)
```
Comment (CodeMyst) [inline Commands/HelpCommand.cs]: Plural of "data" is "data" and not "datas".

File: Commands/HelpCommand.cs [hunk 3]
```diff
@@ -31,22 +34,52 @@ public async Task ModHelp ()
...
+        {
+            var commandList = new List<HelpDataAttribute>();
+
+            foreach (ModuleInfo module in _commands.Modules)
+                foreach (CommandInfo command in module.Commands)
+                {
+                    var data = command.Attributes.FirstOrDefault(a => a is HelpDataAttribute);
+                    if (data != default(HelpDataAttribute))
```
Comment (CodeMyst) [inline Commands/HelpCommand.cs]: Why are you checking if the data (of type `HelpDataAttribute`) has all default values? That data should never have default values, you're already expecting that it has data.

File: Commands/HelpCommand.cs [hunk 4]
```diff
@@ -31,22 +34,52 @@ public async Task ModHelp ()
...
+                    {
+                        var helpData = data as HelpDataAttribute;
```
Comment (CodeMyst) [inline Commands/HelpCommand.cs]: You should've casted that when you were getting the attribute:
```csharp
var data = (HelpDataAttribute) command.Attributes.FirstOrDefault(a => a is HelpDataAttribute);
```

File: Commands/LeaderboardCommand.cs [hunk 5]
```diff
@@ -9,45 +9,40 @@
...
-            TrackedLeaderboards.Add(message.Id, new LeaderboardDisplayData() { DisplayIndex = 0, Message = message, Guild = Context.Guild });
+            _navigator.AddTrackedMessage(new LeaderboardNavigator.LeaderboardDisplayData() { DisplayIndex = 0, Message = message, Guild = Context.Guild });
         }
 
-        private static async Task<EmbedBuilder> BuildForNextPlaces(int startIndex, IGuild guild)
+        private static async Task<EmbedBuilder> BuildForNextPlaces(KarmaTable table, IGuild guild, int startIndex, int pagesize)
         {
-            KarmaTable table = BrackeysBot.Karma;
-            IEnumerable<Tuple<ulong, int>> places = table.GetLeaderboardPlaces(startIndex, LeaderboardPagesize);
+            IEnumerable<Tuple<ulong, int>> places = table.GetLeaderboardPlaces(startIndex, pagesize);
```
Comment (CodeMyst) [inline Commands/LeaderboardCommand.cs]: Use a dictionary here. I'm assuming the `ulong` stands for the user id and `int` for the number of points.

File: Commands/HelpCommand.cs (hunk 3 above)
Comment (YilianSource) [reply Commands/HelpCommand.cs]: My bad for unclear coding. default(HelpDataAttribute) is null, so I am basically checking if the command has a HelpDataAttribute.

File: Commands/LeaderboardCommand.cs (hunk 5 above)
Comment (YilianSource) [reply Commands/LeaderboardCommand.cs]: But you can't order a dictionary.
Even if I were to change it, I would have to change it to IEnumerable<KeyValuePair<ulong, int>> after sorting.

Comment (CodeMyst) [reply Commands/LeaderboardCommand.cs]: Forgo about that. I always use this implementation for a generic ordered dictionary: https://github.com/mattmc3/dotmore/blob/master/dotmore/Collections/Generic/OrderedDictionary.cs

File: Commands/HelpCommand.cs (hunk 3 above)
Comment (CodeMyst) [reply Commands/HelpCommand.cs]: Then it's simpler and more clear to just check if it's null.

**Pull request merged**